### PR TITLE
Runtime changes for IJW Activation

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Add managed entry-points for loading IJW exes and dlls into the runtime.

When resolving tokens in the vtfixup table, call into the ijwhost (if loaded) to resolve delayed-activation thunks back to tokens if needed.

Change the ordering of incremental loading to make DeliverSyncEvents happen before VTableFixups. This change was needed because the IJW assembly needs to be loaded enough to have assemblies it loads assign it as their parent. This requires the IJW assembly to be loaded at least through DeliverSyncEvents when loading dependent assemblies. Dependent assembly loading can happen while running VTableFixups, so either VTableFixups has to happen after DeliverSyncEvents, or the requirement for an assembly to be considered "loaded" (not "active") must be a load level before VTableFixups.

If the IJW assembly is not considered "loaded", then dependent assemblies fall back to loading into the default ALC, which will fail because of the assembly isolation we want for components.

cc: @jeffschwMSFT @sdmaclea 

Contributes to #18013 

See dotnet/core-setup#5185 for the hosting component.

### CODE REVIEW

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentLoadContext.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    internal class ComponentLoadContext : AssemblyLoadContext
```
Comment (vitek-karas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentLoadContext.cs]: I think a better name would be `IsolatedComponentLoadContext` to make it clear that this ALC implementation provides full isolation.

File: src/vm/assemblynative.cpp [hunk 2]
```diff
@@ -372,6 +372,42 @@ void QCALLTYPE AssemblyNative::LoadFromStream(INT_PTR ptrNativeAssemblyLoadConte
...
+    
+    // Need to verify that this is a valid CLR assembly. 
+    if (!pILImage->CheckILFormat())
+        ThrowHR(COR_E_BADIMAGEFORMAT, BFA_BAD_IL);
+    
+    // Get the binder context in which the assembly will be loaded
+    ICLRPrivBinder *pBinderContext = reinterpret_cast<ICLRPrivBinder*>(ptrNativeAssemblyLoadContext);
+    
+    // Pass the stream based assembly as IL and NI in an attempt to bind and load it
+    Assembly* pLoadedAssembly = AssemblyNative::LoadFromPEImage(pBinderContext, pILImage, NULL);
```
Comment (vitek-karas) [inline src/vm/assemblynative.cpp]: The comment says you want to pass the stream to both IL and NI arguments, but the code only passes it to IL.. ?

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            throw new PlatformNotSupportedException();
+#else
+            Assembly entryAssembly = AssemblyLoadContext.Default.LoadFromInMemoryModule(handle);
+            // Emulate traditional app start behavior that adds back on the path of the entry assembly
+            // to the array set for Environment.SetCommandLineArgs.
+            string[] environmentArgs = new string[argv.Length + 1];
+            environmentArgs[0] = entryAssembly.Location;
+            Array.Copy(argv, 0, environmentArgs, 1, argv.Length);
+
+            Environment.SetCommandLineArgs(environmentArgs);
```
Comment (vitek-karas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: It would be much cleaner if we would refactor `CorHost2::ExecuteAssembly` into two parts and the inner section which handles command line arguments and calls execute assembly would be shared between the `CorHost2` method and the new method you're adding for this.
I understand that the end result is the same, but it would be nice to not duplicate code. (It would also be a bit slower due to managed->native->managed marshalling, but the impact should be very small)

File: src/vm/assemblynative.hpp [hunk 4]
```diff
@@ -112,6 +112,8 @@ class AssemblyNative
 
     static BOOL QCALLTYPE GetIsCollectible(QCall::AssemblyHandle pAssembly);
 
+    static INT32 QCALLTYPE ExecuteMainMethod(QCall::AssemblyHandle pAssembly, QCall::ObjectHandleOnStack args);
```
Comment (vitek-karas) [inline src/vm/assemblynative.hpp]: Why not `#ifndef FEATURE_PAL` like for the other one? Because this one works everywhere?

File: src/vm/appdomain.cpp [hunk 5]
```diff
@@ -4313,8 +4313,8 @@ static const char *fileLoadLevelName[] =
     "LOADLIBRARY",                        // FILE_LOAD_LOADLIBRARY
     "POST_LOADLIBRARY",                   // FILE_LOAD_POST_LOADLIBRARY                
     "EAGER_FIXUPS",                       // FILE_LOAD_EAGER_FIXUPS
-    "VTABLE FIXUPS",                      // FILE_LOAD_VTABLE_FIXUPS
     "DELIVER_EVENTS",                     // FILE_LOAD_DELIVER_EVENTS
```
Comment (jkotas) [inline src/vm/appdomain.cpp]: Why this order worked fine for .NET Framework, and it does not work for .NET Core?

File: src/vm/ceeload.cpp [hunk 6]
```diff
@@ -6461,6 +6461,36 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
 }
 
 #if !defined(CROSSGEN_COMPILE)
+using GetTokenForVTableEntry_t = mdToken(STDMETHODCALLTYPE*)(HMODULE module, BYTE**ppVTEntry);
```
Comment (jkotas) [inline src/vm/ceeload.cpp]: Why not just `typedef?

Or is this the new preferred way to say `typedef` in C++ 11 ?

File: src/vm/ceeload.cpp [hunk 7]
```diff
@@ -6461,6 +6461,36 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
...
+        ijwHostModule = CLRGetModuleHandle(W("ijwhost.dll"));
+    }
+
+    return ijwHostModule;
+#else
+    return nullptr;
+#endif // FEATURE_PAL
+}
+
+GetTokenForVTableEntry_t GetTokenGetterFromHostModule()
```
Comment (jkotas) [inline src/vm/ceeload.cpp]: Why do we need this callback when .NET Framework worked fine without it?

File: src/vm/appdomain.cpp (hunk 5 above)
Comment (jkoritzinsky) [reply src/vm/appdomain.cpp]: On .NET Framework, we would correctly load into the same app-domain. For .NET Core, it seems that the an assembly needs to be loaded at least to the FILE_LOAD_DELIVER_EVENTS level for a dependent AssemblySpec to consider it loaded enough to use its ALC to load the assembly.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: > be nice to not duplicate code

FWIW, I do not see much duplication here.

File: src/vm/ceeload.cpp (hunk 6 above)
Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: This is the new way to say `typedef` in C++11. The difference isn't too big, except for function types, where it gets a lot cleaner.

File: src/vm/ceeload.cpp (hunk 7 above)
Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: .NET Framework did all three of these functions (`GetTokenForVTableEntry`, `GetTargetForVTableEntry`, `SetTargetForVTableEntry`) as callbacks to `mscoree.dll`. All three live in the mscoree source.

The reason we need to repeatably check if ijwhost was loaded is because unlike mscoree.dll, ijwhost.dll will ship with the application and may not always be available. Additionally, it allows us to only go to the ijwhost shim when something that uses ijwhost is loaded.

File: src/vm/assemblynative.hpp (hunk 4 above)
Comment (jkoritzinsky) [reply src/vm/assemblynative.hpp]: Yep.

File: src/vm/ceeload.cpp (hunk 7 above)
Comment (jkotas) [reply src/vm/ceeload.cpp]: Is it guaranteed that there is going to be just one ijwhost.dll in the application? I see that you are doing stuff with assembly load contexts...

Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: `ijwhost.dll` will be statically linked to the IJW assembly via the native linker. As far as I understand the Windows loader, if two libraries that each natively statically reference a native dll named `ijwhost.dll` with no path, then after one `ijwhost.dll` is loaded, then later libraries will use the same `ijwhost.dll` module that has been loaded into memory. This loading should not be interfered with because of ALCs AFAIK.

**Comment by jkoritzinsky** [comment]
Any tests we want to add for the native-activation path we'll likely want to do in core-setup since the ijwhost code is the interesting code there. Additionally, those tests will most likely have to wait until we have a compiler that can allow us to link IJW assemblies with ijwhost instead of mscoree.

I have some ideas though on how to test the new System.Private.CoreLib entrypoints within coreclr. I'll get those done next week.

File: src/vm/ceeload.cpp (hunk 7 above)
Comment (jkotas) [reply src/vm/ceeload.cpp]: ok

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    public static class InMemoryAssemblyLoader
+    {
+        public static unsafe int LoadAndExecuteInMemoryAssembly(IntPtr handle, int argc, [MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.LPWStr)] string[] argv)
+        {
+#if FEATURE_PAL
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: I have generally preferred just not building these types when possible. Can we simply exclude this file from build instead?

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/IsolatedComponentLoadContext.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/IsolatedComponentLoadContext.cs]: This was previously a nicely hidden away type in the COM scenario, which is why there wasn't any comments/documentation for it. Now that is is being reused it would be nice for people to know why it was created and when it should be used.

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 10]
```diff
@@ -35,6 +35,11 @@ public abstract class AssemblyLoadContext
         [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
         private static extern IntPtr LoadFromStream(IntPtr ptrNativeAssemblyLoadContext, IntPtr ptrAssemblyArray, int iAssemblyArrayLen, IntPtr ptrSymbols, int iSymbolArrayLen, ObjectHandleOnStack retAssembly);
 
+#if !FEATURE_PAL
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern IntPtr LoadFromInMemoryModule(IntPtr ptrNativeAssemblyLoadContext, IntPtr hModule, ObjectHandleOnStack retAssembly);
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: The P/Invoke and the exposed API shouldn't have the same name. How about naming this one `LoadFromInMemoryModuleInternal`?

File: src/vm/assemblynative.cpp [hunk 11]
```diff
@@ -1331,3 +1367,26 @@ BOOL QCALLTYPE AssemblyNative::InternalTryGetRawMetadata(
...
+
+    BEGIN_QCALL;
+
+    _ASSERTE(pAssembly != nullptr);
+
+    Assembly* assembly = pAssembly->GetAssembly();
+
+    PTRARRAYREF* argsArray = (PTRARRAYREF*)args.m_ppObject;
+
+    exitCode = assembly->ExecuteMainMethod(argsArray, TRUE);
```
Comment (AaronRobinsonMSFT) [inline src/vm/assemblynative.cpp]: Please comment what `TRUE` means.

File: src/vm/assemblynative.cpp [hunk 12]
```diff
@@ -1331,3 +1367,26 @@ BOOL QCALLTYPE AssemblyNative::InternalTryGetRawMetadata(
...
+
+    INT32 exitCode;
+
+    BEGIN_QCALL;
+
+    _ASSERTE(pAssembly != nullptr);
+
+    Assembly* assembly = pAssembly->GetAssembly();
+
+    PTRARRAYREF* argsArray = (PTRARRAYREF*)args.m_ppObject;
```
Comment (AaronRobinsonMSFT) [inline src/vm/assemblynative.cpp]: Is this okay to be null, do we need a check or does `ExecuteMainMethod` handle that?

File: src/vm/ceeload.cpp [hunk 13]
```diff
@@ -6461,6 +6461,36 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
...
+using GetTokenForVTableEntry_t = mdToken(STDMETHODCALLTYPE*)(HMODULE module, BYTE**ppVTEntry);
+
+
+HMODULE GetIJWHostModule()
+{
+#if !defined(FEATURE_PAL)
+    static HMODULE ijwHostModule = nullptr;
+    if (ijwHostModule == nullptr)
+    {
+        ijwHostModule = CLRGetModuleHandle(W("ijwhost.dll"));
```
Comment (AaronRobinsonMSFT) [inline src/vm/ceeload.cpp]: We should talk about this kind of hard-coding. Also, will the `ijwhost` get an add ref?

File: src/vm/ceeload.cpp [hunk 14]
```diff
@@ -6527,6 +6556,13 @@ void Module::FixupVTables()
         return;
     }
 
+    GetTokenForVTableEntry_t GetTokenForVTableEntryCallback = GetTokenGetterFromHostModule();
+
+    if (GetTokenForVTableEntryCallback == nullptr)
```
Comment (AaronRobinsonMSFT) [inline src/vm/ceeload.cpp]: Please comment what is going on here.

File: src/vm/ceeload.cpp (hunk 13 above)
Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: The ijwhost will not get an add-ref with `GetModuleHandle`.

File: src/vm/assemblynative.cpp (hunk 12 above)
Comment (jkoritzinsky) [reply src/vm/assemblynative.cpp]: I've added some null checking in the managed entrypoint to this call.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

File: src/vm/ceeload.cpp (hunk 13 above)
Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: Per our discussion offline, I've re-implemented this method to resolve the module by reading the import tables in the PE and resolving the module handle to whichever module `_CorDllMain` is imported from.

Comment (jkotas) [reply src/vm/ceeload.cpp]: Why is this better?

This looks like load order problems factory - if somebody happens to load desktop IJW assembly first, the pointer will get initialized to point to mscoree and the modules that reference ijwhost and get loaded later will break.

Comment (jkotas) [reply src/vm/ceeload.cpp]: I guess you probably forgot to remove the caching of the entrypoint in the global static from `GetTokenGetterFromHostModule `.

Comment (jkoritzinsky) [reply src/vm/ceeload.cpp]: Yeah I forgot to remove the caching in `GetTokenGetterFromHostModule`. Fixing that now.

**Comment by AaronRobinsonMSFT** [comment]
This PR is looking awfully weird and including all kind of changes. I think you need to rebase your changes on top of what is checked in, rather than merging in.

**Comment by jkoritzinsky** [comment]
I've rebased to clean up the diff.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

**Comment by jkoritzinsky** [comment]
@vitek-karas @AaronRobinsonMSFT can you take another review pass when you have a chance?

File: src/vm/assemblynative.cpp [hunk 15]
```diff
@@ -363,6 +363,42 @@ void QCALLTYPE AssemblyNative::LoadFromStream(INT_PTR ptrNativeAssemblyLoadConte
...
+    BEGIN_QCALL;
+    
+    // Ensure that the invariants are in place
+    _ASSERTE(ptrNativeAssemblyLoadContext != NULL);
+    _ASSERTE(hModule != NULL);
+
+    PEImageHolder pILImage(PEImage::LoadImage((HMODULE)hModule));
+    
+    // Need to verify that this is a valid CLR assembly. 
+    if (!pILImage->CheckILFormat())
```
Comment (AaronRobinsonMSFT) [inline src/vm/assemblynative.cpp]: Is this still needed? Didn't the OS already load the image? I can assure you the OS's PE image verification is far more robust than anything we have.

File: src/vm/assemblynative.cpp [hunk 16]
```diff
@@ -363,6 +363,42 @@ void QCALLTYPE AssemblyNative::LoadFromStream(INT_PTR ptrNativeAssemblyLoadConte
...
+        ThrowHR(COR_E_BADIMAGEFORMAT, BFA_BAD_IL);
+    
+    // Get the binder context in which the assembly will be loaded
+    ICLRPrivBinder *pBinderContext = reinterpret_cast<ICLRPrivBinder*>(ptrNativeAssemblyLoadContext);
+    
+    // Pass the stream based assembly as IL in an attempt to bind and load it
+    Assembly* pLoadedAssembly = AssemblyNative::LoadFromPEImage(pBinderContext, pILImage, NULL); 
+    {
+        GCX_COOP();
+        retLoadedAssembly.Set(pLoadedAssembly->GetExposedObject());
```
Comment (AaronRobinsonMSFT) [inline src/vm/assemblynative.cpp]: How are errors expressed from this function? Are we expecting exceptions or is this `null` on the out?

File: src/vm/ceeload.cpp [hunk 17]
```diff
@@ -6482,6 +6482,63 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
 }
 
 #if !defined(CROSSGEN_COMPILE)
+using GetTokenForVTableEntry_t = mdToken(STDMETHODCALLTYPE*)(HMODULE module, BYTE**ppVTEntry);
+
+HMODULE GetIJWHostForModule(Module* module)
```
Comment (AaronRobinsonMSFT) [inline src/vm/ceeload.cpp]: I am going to assume this function isn't used outside this compilation unit. Please mark it `static` if it is.

File: tests/src/Interop/IJW/ijwhostmock/mscoree.cpp [hunk 18]
```diff
@@ -0,0 +1,38 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <windows.h>
+#include "xplatform.h"
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/IJW/ijwhostmock/mscoree.cpp]: `<xplatform.h>`

File: tests/src/Interop/IJW/ijwhostmock/mscoree.cpp [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+
+#ifndef DLLEXPORT
+#ifdef _MSC_VER
+#define DLLEXPORT __declspec(dllexport)
+#else
+#define DLLEXPORT __attribute__((visibility("default")))
+#endif
+#endif
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/IJW/ijwhostmock/mscoree.cpp]: Perhaps we should add the `DLLEXPORT` macro to `xplatform.h`?

File: src/vm/assemblynative.cpp (hunk 15 above)
Comment (jkoritzinsky) [reply src/vm/assemblynative.cpp]: Good point. I'm going to change this check to just verify that the assembly has the CLR header, which (combined with the OS's PE verification) will verify that the module is a .NET assembly.

File: src/vm/assemblynative.cpp (hunk 16 above)
Comment (jkoritzinsky) [reply src/vm/assemblynative.cpp]: The `AssemblyNative::LoadFromPEImage` will throw an exception if the load fails.

**Comment by jkoritzinsky** [comment]
@AaronRobinsonMSFT @vitek-karas @jkotas anyone have any issues/feedback on what I'm doing in this PR? I'd like to get this merged soon if possible.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/IsolatedComponentLoadContext.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+            string libraryPath = _resolver.ResolveUnmanagedDllToPath(unmanagedDllName);
+            if (libraryPath != null)
+            {
+                return LoadUnmanagedDllFromPath(libraryPath);
+            }
+
+            return IntPtr.Zero;
+        }
+    }
+
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/IsolatedComponentLoadContext.cs]: Remove line.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    public static class InMemoryAssemblyLoader
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: Please document the public type. At least what scenarios this is for.

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs [hunk 22]
```diff
@@ -52,6 +52,34 @@ internal unsafe Assembly InternalLoad(ReadOnlySpan<byte> arrAssembly, ReadOnlySp
 
             return loadedAssembly;
         }
+        
+#if !FEATURE_PAL
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs]: Why is this related to `FEATURE_PAL`? I think @elinor-fung added `PLATFORM_WINDOWS` which seems more appropriate.

Throughout.

File: src/vm/ceeload.cpp [hunk 23]
```diff
@@ -6450,6 +6450,63 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
...
+                        return ijwHost;
+                    }
+
+                }
+            }
+        }
+    }
+
+    return nullptr;
+#else
```
Comment (AaronRobinsonMSFT) [inline src/vm/ceeload.cpp]: Remove the `#else` and always have the `return nullptr;`

File: src/vm/ceeload.cpp [hunk 24]
```diff
@@ -6450,6 +6450,63 @@ void Module::NotifyDebuggerUnload(AppDomain *pDomain)
...
+    return nullptr;
+#endif
+}
+
+static GetTokenForVTableEntry_t GetTokenGetterFromHostModule(HMODULE ijwHost)
+{
+    GetTokenForVTableEntry_t getTokenForVTableEntryWithHost = nullptr;
```
Comment (AaronRobinsonMSFT) [inline src/vm/ceeload.cpp]: Local isn't needed. Just return what is needed.

File: tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def [hunk 25]
```diff
@@ -0,0 +1,3 @@
+EXPORTS
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def]: Can we use `__declspec(dllexport)`?

Comment (jkoritzinsky) [reply tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def]: We need to make sure we match the names on all platforms, and the stdcall mangling breaks that invariant. For the `GetTokenForVTableEntry` entrypoint we could leave it and have CoreCLR look up the mangled name on x86 and the nonmangled name on non-x86 if we want, but we need to have this file anyway for `_CorDllMain` since we need to match the old entrypoint name for our fake mscoree.

Comment (AaronRobinsonMSFT) [reply tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def]: You can control that as follows:

```
#if defined _X86_
#pragma comment(linker, "/export:FuncWithStdCall=_FuncWithStdCall@4")
#endif
```

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs (hunk 22 above)
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs]: It's related to `FEATURE_PAL` because the underlying function in the runtime,  `PEImage::LoadImage(HMODULE)`, is ifdef'd out based on `FEATURE_PAL`. Looking further down the stack, there at the `RawImageLayout` constructor, it looks like the `bTakeOwnership` flag is unimplemented when using the PAL. And then the usage in `GetIJWHostForModule` is for consistency.
https://github.com/dotnet/coreclr/blob/8497ae3ed438f443b22b00a66f42afe8f6623bbf/src/vm/peimagelayout.cpp#L365

File: tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def (hunk 25 above)
Comment (jkoritzinsky) [reply tests/src/Interop/IJW/ijwhostmock/MscoreeExports.def]: I'll do that then.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /// <summary>
+    /// This class enables the .NET IJW host to load an in-memory module as a .NET assembly
+    /// </summary>
+    public static class InMemoryAssemblyLoader
+    {
+        /// <summary>
+        /// Loads into an isolated AssemblyLoadContext an assembly that has already been loaded into memory by the OS loader as a native module.
+        /// </summary>
+        /// <param name="handle">The native handle for the assembly.</param>
```
Comment (vitek-karas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: Nit: I would call this parameter either `module` or `moduleHandle` and the description should state that it's a module handle. In the context of CoreLib/CoreCLR - "native handle" feels like something from CoreCLR - which it's not.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+        /// <param name="assemblyPath">The path to the assembly (as a pointer to a UTF-16 C string).</param>
+        public static unsafe void LoadInMemoryAssembly(IntPtr handle, IntPtr assemblyPath)
+        {
+            AssemblyLoadContext context = new IsolatedComponentLoadContext(Marshal.PtrToStringUni(assemblyPath));
```
Comment (vitek-karas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs]: Can you please add a comment that we intentionally don't cache the ALCs and why (as a contrast to the COM activator which does caching).

File: src/vm/assemblynative.cpp [hunk 28]
```diff
@@ -363,6 +363,42 @@ void QCALLTYPE AssemblyNative::LoadFromStream(INT_PTR ptrNativeAssemblyLoadConte
...
+    PEImageHolder pILImage(PEImage::LoadImage((HMODULE)hModule));
+    
+    // Need to verify that this is a valid CLR assembly. 
+    if (!pILImage->HasCorHeader())
+        ThrowHR(COR_E_BADIMAGEFORMAT, BFA_BAD_IL);
+    
+    // Get the binder context in which the assembly will be loaded
+    ICLRPrivBinder *pBinderContext = reinterpret_cast<ICLRPrivBinder*>(ptrNativeAssemblyLoadContext);
+    
+    // Pass the stream based assembly as IL in an attempt to bind and load it
```
Comment (vitek-karas) [inline src/vm/assemblynative.cpp]: This is a copy/paste from the stream case - in this one there are no streams - so the comment is misleading.

**Comment by jkoritzinsky** [comment]
The two failures are infra failures due to their flow jobs still being created even after being deleted. So I'm going to merge this in since everything else passed.
